cookbooks/prescribed_velocity_ascii_data/prescribed_velocity_ascii_data.cc

```diff
@@ -0,0 +1,350 @@
+                      );
+
+    prm.enter_subsection ("Prescribed velocities");
+    {
+      {
+        if (dim == 2)
+          {
+            prescribed_velocity_field_params_2d.declare_parameters(prm, "blob", "blob", "Ascii data model");
+            //prescribed_velocity_field_data_2d.initialize(1);
+          }
+        else
+          {
+            prescribed_velocity_field_params_3d.declare_parameters(prm, "blob", "blob", "Ascii data model");
+            //prescribed_velocity_field_data_3d.initialize(1);
+          }
+      }
+
+      prm.enter_subsection ("Velocity function");
+      {
+        if (dim == 2)
+          Functions::ParsedFunction<2>::declare_parameters (prm, 2);
+        else
+          Functions::ParsedFunction<3>::declare_parameters (prm, 3);
+      }
+      prm.leave_subsection ();
+    }
+    prm.leave_subsection ();
+  }
+
+  template <int dim>
+  void parse_parameters(const Parameters<dim> &,
+                        ParameterHandler &prm)
+  {
+    prescribe_internal_velocities = prm.get_bool ("Prescribe internal velocities");
+    prm.enter_subsection ("Prescribed velocities");
+    {
+      {
+        if (dim == 2)
+          {
+            prescribed_velocity_field_params_2d.parse_parameters(prm, "Ascii data model");
+          }
+        else
+          {
+            prescribed_velocity_field_params_3d.parse_parameters(prm, "Ascii data model");
+          }
+      }
+
+      prm.enter_subsection("Velocity function");
+      {
+        try
+          {
+            if (dim == 2)
+              prescribed_velocity_function_2d.parse_parameters (prm);
+            else
+              prescribed_velocity_function_3d.parse_parameters (prm);
+          }
+        catch (...)
+          {
+            std::cerr << "ERROR: FunctionParser failed to parse\n"
+                      << "\t'Prescribed velocities.Velocity function'\n"
+                      << "with expression\n"
+                      << "\t'" << prm.get("Function expression") << "'";
+            throw;
+          }
+      }
+      prm.leave_subsection();
+    }
+    prm.leave_subsection ();
+
+    prescribed_velocity_ascii_file_loaded = false;
+  }
+
+
+
+  /**
+   * This function retrieves the unit support points (in the unit cell) for the current element.
+   * The DGP element used when 'set Use locally conservative discretization = true' does not
+   * have support points. If these elements are in use, a fictitious support point at the cell
+   * center is returned for each shape function that corresponds to the pressure variable,
+   * whereas the support points for the velocity are correct; the fictitious points don't matter
+   * because we only use this function when interpolating the velocity variable, and ignore the
+   * evaluation at the pressure support points.
+   */
+  template <int dim>
+  std::vector< Point<dim> >
+  get_unit_support_points_for_velocity(const SimulatorAccess<dim> &simulator_access)
```

Comment:
This is a tool function copied somewhere from within ASPECT, right? If so, maybe we could move that function into namespace `Utilities` so that it doesn't have to be duplicated.

Reply:
It's copied from @jperryhouts' cookbook and from the associated test, so it isn't used anywhere in ASPECT's source directory. 

Do you want it moved, @bangerth? I'm happy to do that if you think it'll be more widely used.